project/src/behaviors/GSTBlend.gd

```diff
@@ -0,0 +1,53 @@
+extends GSTSteeringBehavior
+class_name GSTBlend
+"""
+Blends multiple steering behaviors into one, and returns acceleration combining all of them.
+
+Each behavior is associated with a weight - a modifier by which the result will be multiplied by,
+then added to a total acceleration.
+"""
+
+
+onready var _behaviors: = []
+onready var _accel: = GSTTargetAcceleration.new()
+
+
+func _init(agent: GSTSteeringAgent).(agent) -> void:
+	pass
+
+
+func add(behavior: GSTSteeringBehavior, weight: float) -> void:
+	behavior.agent = agent
+	_behaviors.append(GSTBehaviorAndWeight.new(behavior, weight))
+
+
+func get_behavior_at(index: int) -> GSTBehaviorAndWeight:
+	if _behaviors.size() > index:
+		return _behaviors[index]
+	printerr("Tried to get index " + str(index) + " in array of size " + str(_behaviors.size()))
+	return null
+
+
+func _calculate_steering(blended_accel: GSTTargetAcceleration) -> GSTTargetAcceleration:
+	blended_accel.set_zero()
+	
+	for i in range(_behaviors.size()):
+		var bw: GSTBehaviorAndWeight = _behaviors[i]
+		bw.behavior.calculate_steering(_accel)
+		
+		blended_accel.add_scaled_accel(_accel, bw.weight)
+	
+	blended_accel.linear = Utils.clampedv3(blended_accel.linear, agent.max_linear_acceleration)
+	if blended_accel.angular > agent.max_angular_acceleration:
+		blended_accel.angular = agent.max_angular_acceleration
+	
+	return blended_accel
+
+
+class GSTBehaviorAndWeight:
```

Comment:
For internal data (with no behavior) let's consider the pros and cons of using `dictionaries`.  Array of `GSTBehaviorAndWeigth` vs Array of `{behavior = GSTSteeringBehavior, weight = 1.0}`.

As we noted before object creation is very costly so what about `dictionaries` then? Which are supposed to be super-light. Yeah, the downside is we won't have auto-completion and we loose a bit of clarity, but I think it can be regained through good docs/comments. These are supposed to be for internal use anyway, right?

Any thoughts?

Reply:
Would probably be lighter memory footprint and easier on performance to have an array of dictionaries. Though the opaqueness and lack of type safety is the main reason I didn't go for it. Internally, it's not a big deal, I can cast.

But I don't want to remove the ability for the user to change weights or access one of the behaviors through the blend. So we'll definitely need to be clear about the contents of the dictionary when they get it, but so long as it's clearly noted, shouldn't be that big a deal.